scalems/wrappers/__init__.py

```diff
@@ -16,32 +15,109 @@
+        OpenMM: (look at ExTasy workflows)
+        PLUMED-GROMACS: GROMACS + PLUMED input file
+        LAMMPS: .input file
+
+        Crucially, it provide a place for the information to be staged
+        for modifying input, such that a single simulation can be
+        loaded, and then variants of that simulation can be "cloned out", and modified.
+
+        For many programs, this may simply be putting the text of the
+        files into a container where they can be retrived and
+        modified.
+
     modify_input(simulation_input: SimulationInput, **kwargs) -> SimulationInput:
+
         A simulation input data source produced by *make_input* or *modify_input*
         may be modified in a pipe line through (implementation dependent) key word
         arguments to a *modify_input* command.
 
+        For most options, then a keyword argument can directly
+        parallel the keyword syntax used by the simulation programs
+        themselves: LAMMPS, PLUMED, GROMACS (in the .mdp) all use
+        keyword options.  It's not as clear for OpenMM, but we should
+        use the ExTASY implementations to guide us as to how modification is done there.  
+
+        One question is how the defintion of the chemical system
+        should changed using this approach. This is not set up very
+        well in the keyword format.  We could, for now, just replace
+        the file containing the molecular system definition entirely
+        using a modify_input.  This might be the best bet for now, but
+        we will want to think about what sorts of changes might be of
+        interest, if any in the future.
+
+        The issue is that in most cases, if the chemical system is
+        changing, it usually requires changes of many lines of
+        topologies, and so passing it in as a dictionary will likely
+        be a big pain. Likely easier to take care of it outside the
+        simulation for now. Eventually, if some parallel work on
+        simulation topology definitions in Open Force Field works,
+        then it may be possible to do this better.
+
+        Things like "training potential" that get attached in the
+        examples should probably be attached at this point in the
+        workflow, rather than at the simulate. For most workflows, it
+        could ONLY be added at this stage, since potentials are
+        defined in the input files.  For GMX_API driven simulations,
+        the actual attachment of the potentials to the md simulation
+        could happen in the simulate function, but it would be better
+        if they were loaded into the data structure in the
+        modify_input sage.
+
     simulate(simulation_input: SimulationInput, **kwargs) -> Simulation:
         Create a simulation operation instance. Accepts input from *make_input*
         or *modify_input*, along with key word arguments specified by the wrapper
         module.
 
+        Potentially, we use this to specify which resources assoicated
```

Comment:
Resources should not be specified by the user, unless we're using "resources" differently.  If there are resource-type requirements or preferences (e.g. GPU), those could be attached at some point, but I'm not sure if this is the right place.

Reply:
There is currently a bit of an outstanding question on how/where to specify run-time parameters, especially when they are task-specific.

I think that's somewhat of a separate issue, but we actually need a general solution for describing the resource requirements or expectations of individual tasks, both as a necessary implementation detail (e.g. can/should a given executable by launched with `mpiexec`) and to allow hints to the scheduler (please run task X in a single process, give task Y as many processes as you feel like, and make sure task Z occupies exactly one node because its author wrote it that way, or tell task W to use 1 28-thread process, because I happen to know that's the type of node it will get launched on).

But part of the "wrappers" documentation should define the line between the specified abstraction and the implementation-specific detail.

We could identify a particular function signature that is allowed to accept `**kwargs` (or a specific key word) to accommodate tools that need to be told how to use computing resources. Some other resource specification might be through module variables, config files, or separate instructions to a manager for the execution environment (e.g. like the `with context.launch(workflow, execution_parameters) as session: session.run()` sorts of constructs we used in gmxapi <0.1).

Reply:
> Resources should not be specified by the user, unless we're using "resources" differently.

This is a good point about definitions.  Maybe "resources requested" by each the simulation?  I'm thinking about the sorts of things that go into the mdrun invocation (like -npme, -nt, etc).  I think that the user does have to specify if they want a simulation to use a large amount of resources per simulation or small number of resources per simulation.  Additionally, on complex computational resources like Summit, etc. then the chance that an automated procedure could guess the right flags is virtually nil. (we've just been though this optimizing running GROMACS on Summit).  Thus, I think the right way to handle this in the near term is to have each unit of simulation specify what the user would like it use, and the scheduler part is in charge of funding those resources and placing/scheduling the simulation.

I can imagine a further development where maybe the scheduler can override some of these, but I think the potential for things to go badly wrong is bit. I would be more in favor of just giving up (logging the reasons why) if it's not possible for the users's request to be satisfied.  

> If there are resource-type requirements or preferences (e.g. GPU), those could be attached at some point, but I'm not sure if this is the right place.

My understanding is that the 'simulate' invocation (at least the way Eric and I were discussing it) is the last place the user touches the task before it's handed over to the scheduler, so it seems like the last possible time to attach that information.